### PR TITLE
Add dns_kinghost.sh

### PR DESCRIPTION
**Pull request by fbraz3**
kinghost.com.br is one of the biggest registrars/web hosts in Brazil and the very first on this country to officially support letsencrypt! =)

### CODE REVIEW

**Comment by Neilpang** [comment]
check this guide.
https://github.com/Neilpang/acme.sh/wiki/DNS-API-Dev-Guide

File: dnsapi/dns_kinghost.sh [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+KING_Api="https://api.kinghost.net/acme"
+
+# Usage: add  _acme-challenge.www.domain.com   "XKrxpRBosdIKFzxW_CT3KLZNf6q0HG9i01zxXp5CPBs"
+# Used to add txt record
+dns_kinghost_add() {
+    fulldomain=$1
+    txtvalue=$2
+
+    KINGHOST_username="${KINGHOST_username:-$(_readaccountconf_mutable KINGHOST_username)}"
```
Comment (Neilpang) [inline dnsapi/dns_kinghost.sh]: rename:  `KINGHOST_username` => `KINGHOST_Username`

File: dnsapi/dns_kinghost.sh [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+    #save the credentials to the account conf file.
+    _saveaccountconf_mutable KINGHOST_username  "$KINGHOST_username"
+    _saveaccountconf_mutable KINGHOST_Password  "$KINGHOST_Password"
+
+    _debug "Getting txt records"
+    _kinghost_rest GET "dns" "name=$fulldomain&content=$txtvalue"
+
+    #This API call returns "status":"ok" if dns record does not exists
+    #We are creating a new txt record here, so we expect the "ok" status
+    if ! printf "%s" "$response" | grep '"status":"ok"' >/dev/null; then
```
Comment (Neilpang) [inline dnsapi/dns_kinghost.sh]: use `echo`  instead of `printf`

File: dnsapi/dns_kinghost.sh (hunk 1 above)
Comment (fbraz3) [reply dnsapi/dns_kinghost.sh]: My bad!

Changed as required. ;)

File: dnsapi/dns_kinghost.sh (hunk 2 above)
Comment (fbraz3) [reply dnsapi/dns_kinghost.sh]: changed!

**Comment by martgras** [comment]
please fix the travis errors. 
The travis checks are failing because you're using indentation with 4 spaces instead of 2.

**Comment by fbraz3** [comment]
Hi @martgras 
I've fixed the identation as requested, but it's still failing!
I can't find what's wrong now. =(

File: dnsapi/README.md [hunk 3]
```diff
@@ -787,6 +787,17 @@ acme.sh --issue --dns dns_da -d example.com -d www.example.com
 
 The `DA_Api` and `DA_Api_Insecure` will be saved in `~/.acme.sh/account.conf` and will be reused when needed.
 
+## 42. Use KingHost DNS API
+
+API access must be enabled at https://painel.kinghost.com.br/painel.api.php
+
+```
+export KINGHOST_username="yourusername"
```
Comment (Neilpang) [inline dnsapi/README.md]: KINGHOST_Username

File: dnsapi/dns_kinghost.sh [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+  if [ -z "$KINGHOST_Username" ] || [ -z "$KINGHOST_Password" ]; then
+    KINGHOST_Username=""
+    KINGHOST_Password=""
+    _err "You don't specify KingHost api password and email yet."
+    _err "Please create you key and try again."
+    return 1
+  fi
+
+  #save the credentials to the account conf file.
+  _saveaccountconf_mutable KINGHOST_Username  "$KINGHOST_Username"
```
Comment (Neilpang) [inline dnsapi/dns_kinghost.sh]: replace the 2 spaces with only one space here.

File: dnsapi/dns_kinghost.sh [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+    response="$(_get "$KING_Api/$uri.json?$data")"
+  fi
+
+  if [ "$?" != "0" ]; then
+    _err "error $uri"
+    return 1
+  fi
+  _debug2 response "$response"
+  return 0
+}
```
Comment (Neilpang) [inline dnsapi/dns_kinghost.sh]: add a new empty line at the end of file

File: dnsapi/dns_kinghost.sh [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+  fi
+
+  _kinghost_rest POST "dns" "name=$fulldomain&content=$txtvalue"
+  if ! echo "$response" | grep '"status":"ok"' >/dev/null; then
+    _err "Error"
+    _err "$response"
+    return 1
+  fi
+
+  return 0;
```
Comment (Neilpang) [inline dnsapi/dns_kinghost.sh]: remove `;`  here

File: dnsapi/dns_kinghost.sh [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+
+# Usage: fulldomain txtvalue
+# Used to remove the txt record after validation
+dns_kinghost_rm() {
+  fulldomain=$1
+  txtvalue=$2
+
+  KINGHOST_Password="${KINGHOST_Password:-$(_readaccountconf_mutable KINGHOST_Password)}"
+  KINGHOST_Username="${KINGHOST_Username:-$(_readaccountconf_mutable KINGHOST_Username)}"
+    if [ -z "$KINGHOST_Password" ] || [ -z "$KINGHOST_Username" ]; then
```
Comment (Neilpang) [inline dnsapi/dns_kinghost.sh]: reduce the indent.

**Comment by Neilpang** [comment]
why not check the CI errors here: https://travis-ci.org/Neilpang/acme.sh/jobs/358428399

**Comment by fbraz3** [comment]
It's failing at this line
`The command "if [ "$TRAVIS_OS_NAME" = "linux" ]; then git diff --exit-code && echo "shfmt OK" ; fi" exited with 1.`

But on my console returns ok
`felipebraz@CW000685:~/PhpstormProjects/acme.sh$  git diff --exit-code && echo "shfmt OK"
shfmt OK
`

**Comment by fbraz3** [comment]
i don't knew shfmt utility until now. Installing and testing here

**Comment by Neilpang** [comment]
Hi,

what is  dns_infraws.sh in your PR ?

**Comment by fbraz3** [comment]
Hi @Neilpang 
dns_infraws.sh was a internal test, was pushed by mistake. Removed!

**Pull request merged**